server/polar/subscription/service.py

```diff
@@ -568,51 +591,427 @@ async def update_subscription_from_stripe(
+            subscription.canceled_at = canceled_at
 
     async def _after_subscription_updated(
         self,
         session: AsyncSession,
         subscription: Subscription,
+        *,
         previous_status: SubscriptionStatus,
-        previous_cancel_at_period_end: bool,
+        previous_ends_at: datetime | None,
     ) -> None:
         # Webhooks
+        await self._on_subscription_updated(session, subscription)
+
+        became_activated = subscription.active and not SubscriptionStatus.is_active(
+            previous_status
+        )
+        if became_activated:
+            await self._on_subscription_activated(session, subscription)
+
+        is_canceled = subscription.ends_at and subscription.canceled_at
+        updated_ends_at = subscription.ends_at != previous_ends_at
+        cancellation_changed = is_canceled and updated_ends_at
+        if cancellation_changed:
+            await self._on_subscription_canceled(session, subscription)
+
+        became_uncanceled = previous_ends_at and not is_canceled
+        if became_uncanceled:
+            await self._on_subscription_uncanceled(session, subscription)
+
+        became_revoked = subscription.revoked and not SubscriptionStatus.is_revoked(
+            previous_status
+        )
+        if became_revoked:
+            await self._on_subscription_revoked(session, subscription)
+
+    async def _on_subscription_updated(
+        self,
+        session: AsyncSession,
+        subscription: Subscription,
+    ) -> None:
         await self._send_webhook(
             session, subscription, WebhookEventType.subscription_updated
         )
-        if subscription.active and not SubscriptionStatus.is_active(previous_status):
-            await self._send_webhook(
-                session, subscription, WebhookEventType.subscription_active
-            )
-        if subscription.revoked and not SubscriptionStatus.is_revoked(previous_status):
-            await self._send_webhook(
-                session, subscription, WebhookEventType.subscription_revoked
-            )
-        if subscription.cancel_at_period_end and not previous_cancel_at_period_end:
-            await self._send_webhook(
-                session, subscription, WebhookEventType.subscription_canceled
-            )
 
-        # Notifications
-        if subscription.active and SubscriptionStatus.is_incomplete(previous_status):
-            await self._send_new_subscription_notification(session, subscription)
+    async def _on_subscription_activated(
+        self,
+        session: AsyncSession,
+        subscription: Subscription,
+    ) -> None:
+        await self._send_webhook(
+            session, subscription, WebhookEventType.subscription_active
+        )
+
+        # TODO: Copied this logic over. Why the extra check for customer
+        # confirmation vs. webhook/creator notification? Look into it, but avoid
+        # changing for now.
```

Comment:
This logic was triggered each time we received an update from Stripe. So this check was there to avoid sending a confirmation email at each billing cycle 🙃

But since here it's wrapped in a `_on_subscription_activated` method where we already know the subscription is becoming active for the first time, then it's safe to remove it.